### PR TITLE
Merge clause PR

### PR DESCRIPTION
**Pull request by DvirDukhan**
@hvub 
Merge clause PR
High-level semantics:
Merge1 to Merge4 - Merge over nodes (Merge, Merge On Create, Merge On Match, Merge On Create and On Match)
Merge5 to Merge8 - Merge over edges (Merge, Merge On Create, Merge On Match, Merge On Create and On Match)
Precategorized - Interop with other clauses
Thanks

### CODE REVIEW

File: tck/features/clauses/merge/Merge1.feature [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+      WITH 42 AS var
+      MERGE (c:N {var: var})
+      """
+    Then the result should be empty
+    And the side effects should be:
+      | +nodes      | 1 |
+      | +labels     | 1 |
+      | +properties | 1 |
+
+  Scenario: Should support updates while merging
```
Comment (DvirDukhan) [inline tck/features/clauses/merge/Merge1.feature:159]:
```suggestion
  Scenario: Merge should support updates while merging
```

File: tck/features/clauses/merge/Merge1.feature [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+      """
+    Then the result should be, in any order:
+      | labels     |
+      | ['L', 'B'] |
+    And the side effects should be:
+      | +nodes      | 1 |
+      | +labels     | 1 |
+      | +properties | 1 |
+
+  Scenario: Introduce named paths 2
```
Comment (DvirDukhan) [inline tck/features/clauses/merge/Merge1.feature:211]:
```suggestion
  Scenario: Merge a node in a named path
```

File: tck/features/clauses/merge/Merge3.feature [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+# proposals for change that have been documented or implemented should only be
+# described as "implementation extensions to Cypher" or as "proposed changes to
+# Cypher that are not yet approved by the openCypher community".
+#
+
+#encoding: utf-8
+
+Feature: Merge3 - Merge Node - On Match
+
+  Scenario: Should be able to set labels on match
```
Comment (DvirDukhan) [inline tck/features/clauses/merge/Merge3.feature:33]:
```suggestion
  Scenario: Merge should be able to set labels on match
```

File: tck/features/clauses/merge/Merge4.feature [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+# proposals for change that have been documented or implemented should only be
+# described as "implementation extensions to Cypher" or as "proposed changes to
+# Cypher that are not yet approved by the openCypher community".
+#
+
+#encoding: utf-8
+
+Feature: Merge4 - Merge Node - On Match and On Create
+
+  Scenario: Should be able to set labels on match and on create
```
Comment (DvirDukhan) [inline tck/features/clauses/merge/Merge4.feature:33]:
```suggestion
  Scenario: Merge should be able to set labels on match and on create
```

File: tck/features/precategorized/clauses/merge/Merge1-1.feature [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+# proposals for change that have been documented or implemented should only be
+# described as "implementation extensions to Cypher" or as "proposed changes to
+# Cypher that are not yet approved by the openCypher community".
+#
+
+#encoding: utf-8
+
+Feature: Merge1-1 - Merge Node and Match Interoperability
+
+  Scenario: Should be able to merge using property from match
```
Comment (DvirDukhan) [inline tck/features/precategorized/clauses/merge/Merge1-1.feature:33]:
```suggestion
  Scenario: Merge should be able to merge using property from match
```

File: tck/features/precategorized/clauses/merge/Merge1-1.feature [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+      MATCH (person:Person)
+      MERGE (city:City {name: person.bornIn})
+      """
+    Then the result should be empty
+    And the side effects should be:
+      | +nodes      | 3 |
+      | +labels     | 1 |
+      | +properties | 3 |
+
+  Scenario: Should be able to use properties from match in ON MATCH
```
Comment (DvirDukhan) [inline tck/features/precategorized/clauses/merge/Merge1-1.feature:55]:
```suggestion
  Scenario: Merge should be able to use properties from match in ON MATCH
```

File: tck/features/precategorized/clauses/merge/Merge1-1.feature [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    Then the result should be, in any order:
+      | person.bornIn |
+      | 'New York'    |
+      | 'Ohio'        |
+    And the side effects should be:
+      | +nodes      | 1 |
+      | +labels     | 1 |
+      | +properties | 1 |
+
+  Scenario: Should be able to use properties from match in ON MATCH and ON CREATE
```
Comment (DvirDukhan) [inline tck/features/precategorized/clauses/merge/Merge1-1.feature:78]:
```suggestion
  Scenario: Merge should be able to use properties from match in ON MATCH and ON CREATE
```

File: tck/features/precategorized/clauses/merge/Merge1-2.feature [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+# proposals for change that have been documented or implemented should only be
+# described as "implementation extensions to Cypher" or as "proposed changes to
+# Cypher that are not yet approved by the openCypher community".
+#
+
+#encoding: utf-8
+
+Feature: Merge1-2 - Merge Node and Create Interoperability
+
+  Scenario: Should handle arguments properly with only write clauses
```
Comment (DvirDukhan) [inline tck/features/precategorized/clauses/merge/Merge1-2.feature:33]:
```suggestion
  Scenario: Merge should handle arguments properly with only write clauses
```

File: tck/features/precategorized/clauses/merge/Merge1-2.feature [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+      """
+      CREATE (a {num: 1})
+      MERGE ({v: a.num})
+      """
+    Then the result should be empty
+    And the side effects should be:
+      | +nodes      | 2 |
+      | +properties | 2 |
+
+  Scenario: Should be able to use properties from match in ON CREATE
```
Comment (DvirDukhan) [inline tck/features/precategorized/clauses/merge/Merge1-2.feature:45]:
```suggestion
  Scenario: Merge should be able to use properties from match in ON CREATE
```

File: tck/features/clauses/merge/Merge5.feature [hunk 10]
```diff
@@ -290,23 +178,6 @@ Feature: MergeRelationshipAcceptance
...
-      """
-    Then the result should be, in any order:
-      | count(*) |
-      | 1        |
-    And the side effects should be:
-      | +nodes         | 3 |
-      | +relationships | 2 |
-      | +labels        | 3 |
-
   Scenario: Introduce named paths 1
```
Comment (DvirDukhan) [inline tck/features/clauses/merge/Merge5.feature:181]:
```suggestion
  Scenario: Merge entities in named path
```

**Review by DvirDukhan — COMMENTED** [review]
@hvub
here are some renaming suggestions for this PR

**Review by hvub — CHANGES_REQUESTED** [review]
Generally this looks pretty good, I think. I have made a couple of suggestions.

Mainly, I suggest to be a little more proactive with scenarios that combine `MERGE` with other clauses. It is not necessarily exactly what we did so far, but maybe a semi-consistent concept of how deal with these interoperability scenarios can arise from that in the following weeks.

Specifically my thinking is, scenarios that test whether `MERGE` does or does not do something based variables that other clauses have bound or delete before seems to be clearly testing functionality of `MERGE`.

I have not made similar suggestion for `UNWIND` before `MERGE` – I am/was still unsure about it – although, I now realize, it fits the description given about. Probably, we should be adventurous and move these `UNWIND` before `MERGE` scenarios into the appropriate `MergeX.feature` files, as well.

I saw one scenario that does `CREATE` + `MERGE` + `CREATE`. That is less clear, so I am pretty sure it should stay in pre-categorized for now.

@DvirDukhan If you could make the adjustments and add the numbering, please. I will then have a final look. Thx.

File: tck/features/precategorized/clauses/merge/Merge2-2.feature [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+# proposals for change that have been documented or implemented should only be
+# described as "implementation extensions to Cypher" or as "proposed changes to
+# Cypher that are not yet approved by the openCypher community".
+#
+
+#encoding: utf-8
+
+Feature: Merge2-2 - Merge Edge and Unwind Interoperability
+
+  Scenario: [1]UNWIND with multiple merges
```
Comment (hvub) [inline tck/features/precategorized/clauses/merge/Merge2-2.feature:33]:
```suggestion
  Scenario: [1] UNWIND with multiple merges
```
space missing.

**Review by hvub — APPROVED** [review]
Assuming the inclusion of the little suggestion about the missing space, this look good and can be merged.

**Pull request merged**